openhands/core/message.py

```diff
@@ -60,60 +57,24 @@ def contains_image(self) -> bool:
 
     @model_serializer
     def serialize_model(self) -> dict:
-        content: list[dict[str, str | dict[str, str]]] = []
-
-        for item in self.content:
-            if isinstance(item, TextContent):
-                content.append(item.model_dump())
-            elif isinstance(item, ImageContent):
-                content.extend(item.model_dump())
-
-        return {'content': content, 'role': self.role}
-
-
-def format_messages(
-    messages: Union[Message, list[Message]],
-    with_images: bool,
-    with_prompt_caching: bool,
-) -> list[dict]:
-    if not isinstance(messages, list):
-        messages = [messages]
-
-    if with_images or with_prompt_caching:
-        return [message.model_dump() for message in messages]
-
-    converted_messages = []
-    for message in messages:
-        content_parts = []
-        role = 'user'
-
-        if isinstance(message, str) and message:
-            content_parts.append(message)
-        elif isinstance(message, dict):
-            role = message.get('role', 'user')
-            if 'content' in message and message['content']:
-                content_parts.append(message['content'])
-        elif isinstance(message, Message):
-            role = message.role
-            for content in message.content:
-                if isinstance(content, list):
-                    for item in content:
-                        if isinstance(item, TextContent) and item.text:
-                            content_parts.append(item.text)
-                elif isinstance(content, TextContent) and content.text:
-                    content_parts.append(content.text)
-        else:
-            logger.error(
-                f'>>> `message` is not a string, dict, or Message: {type(message)}'
+        content: list[dict] | str
+        if self.role == 'system':
+            # For system role, concatenate all text content into a single string
+            content = '\n'.join(
+                item.text for item in self.content if isinstance(item, TextContent)
             )
-
-        if content_parts:
-            content_str = '\n'.join(content_parts)
-            converted_messages.append(
-                {
-                    'role': role,
-                    'content': content_str,
-                }
+        elif self.role == 'assistant' and not self.contains_image:
```

Comment:
nit: this branch seems mergeable with previous branch

Reply:
You're right, I would suggest to keep it at least for now though. The branching helps my brain to see how we serialize each of the 3 roles, and I suspect we will continue to need to look / tweak this code...